config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java

```diff
@@ -689,6 +747,98 @@ protected void configure(HttpSecurity http) throws Exception {
 		}
 	}
 
+	@EnableWebSecurity
+	static class CustomJwtDecoderOnDsl extends WebSecurityConfigurerAdapter {
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			// @formatter:off
+			http
+				.authorizeRequests()
+					.anyRequest().authenticated()
+					.and()
+				.oauth2()
+					.resourceServer()
+						.jwt()
+							.decoder(new CustomJwtDecoder());
+			// @formatter:on
+		}
+	}
+
+	@EnableWebSecurity
+	static class CustomJwtDecoderAsBean extends WebSecurityConfigurerAdapter {
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			// @formatter:off
+			http
+				.authorizeRequests()
+					.anyRequest().authenticated()
+					.and()
+				.oauth2()
+					.resourceServer();
+			// @formatter:on
+		}
+
+		@Bean
+		public JwtDecoder decoder() {
+			return new CustomJwtDecoder();
+		}
+	}
+
+	@EnableWebSecurity
+	static class CustomJwtDecoderConflicting extends WebSecurityConfigurerAdapter {
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			// @formatter:off
+			http
+				.authorizeRequests()
+					.anyRequest().authenticated()
+					.and()
+				.oauth2()
+					.resourceServer()
+						.jwt()
+							.decoder(new CustomJwtDecoder());
+			// @formatter:on
+		}
+
+		@Bean
+		public JwtDecoder decoder() {
+			return new NimbusJwtDecoderJwkSupport("https://example.org");
+		}
+	}
+
+	@EnableWebSecurity
+	static class CustomJwtDecoderAndJwkSetUri extends WebSecurityConfigurerAdapter {
+		@Value("${mock.jwk-set-uri:https://example.org}") String uri;
+
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			// @formatter:off
+			http
+				.authorizeRequests()
+					.anyRequest().authenticated()
+					.and()
+				.oauth2()
+					.resourceServer()
+						.jwt()
+							.jwkSetUri(this.uri);
+			// @formatter:on
+		}
+
+		@Bean
+		public JwtDecoder decoder() {
+			return new CustomJwtDecoder();
+		}
+	}
+
+	static class CustomJwtDecoder implements JwtDecoder {
```

Comment:
I'd prefer we just use a mock instead of a custom implementation. This is why Mockito exists and it helps reduces boilerplate code.